### PR TITLE
feat: Bidirectional Streaming for User Defined Source

### PR DESCRIPTION
**Pull request by yhl25**

### CODE REVIEW

**Comment by yhl25** [comment]
I have updated the udsource e2e tests to use docker images from my personal directory, so that the tests don't fail. Once the main repo images are updated to use streaming source I will revert those changes.

File: pkg/sources/udsource/grpc_udsource.go [hunk 1]
```diff
@@ -175,13 +175,20 @@ func (u *GRPCBasedUDSource) ApplyAckFn(ctx context.Context, offsets []isb.Offset
...
+			Request: &sourcepb.AckRequest_Request{
+				Offset: offset,
+			},
+		}
+		ackRequests[i] = r
+	}
+	_, err := u.client.AckFn(ctx, ackRequests)
+	if err != nil {
+		return err
 	}
```
Comment (KeranYang) [inline pkg/sources/udsource/grpc_udsource.go:190]: nit: can directly return err.

File: test/udsource-e2e/testdata/simple-source-rust.yaml [hunk 2]
```diff
@@ -0,0 +1,28 @@
...
+        udsource:
+          container:
+            # A simple user-defined source for e2e testing
+            # See https://github.com/numaproj/numaflow-go/tree/main/pkg/sourcer/examples/simple_source
+            image: quay.io/yhl25/numaflow-rs/simple-source:stable
+      limits:
+        readBatchSize: 500
+      scale:
+        # set it as two pods to be different from the sink such that we can use this pipeline
+        # to test processing rate is consistent across vertices when they have different replica counts.
```
Comment (KeranYang) [inline test/udsource-e2e/testdata/simple-source-rust.yaml:18]:
```suggestion
```

File: rust/monovertex/Cargo.toml [hunk 3]
```diff
@@ -34,10 +34,11 @@ backoff = { path = "../backoff" }
 parking_lot = "0.12.3"
 prometheus-client = "0.22.3"
 kube = "0.94.0"
+log = "0.4.22"
 
 [dev-dependencies]
 tempfile = "3.11.0"
-numaflow = { git = "https://github.com/numaproj/numaflow-rs.git", branch = "main" }
+numaflow = { git = "https://github.com/numaproj/numaflow-rs.git", branch = "handshake" }
```
Comment (vigith) [inline rust/monovertex/Cargo.toml:41]:
```suggestion
numaflow = { git = "https://github.com/numaproj/numaflow-rs.git", branch = "main" }
```

eventually we should point to a stable release in crates.io

File: test/e2e/functional_test.go [hunk 4]
```diff
@@ -41,7 +41,7 @@ func (s *FunctionalSuite) TestCreateSimplePipeline() {
 	w := s.Given().Pipeline("@testdata/simple-pipeline.yaml").
 		When().
 		CreatePipelineAndWait()
-	defer w.DeletePipelineAndWait()
+	//defer w.DeletePipelineAndWait()
```
Comment (whynowy) [inline test/e2e/functional_test.go:44]: Uncomment it?

**Review by vigith — APPROVED** [review]

**Pull request merged**